core/lib/multivm/src/versions/testonly/account_validation_rules.rs

```diff
@@ -0,0 +1,109 @@
+use assert_matches::assert_matches;
+use ethabi::Token;
+use zksync_types::{
+    fee::Fee, l2::L2Tx, transaction_request::TransactionRequest, Address, Eip712Domain, Execute,
+    L2ChainId, U256,
+};
+use zksync_vm_interface::{tracer::ViolatedValidationRule, InspectExecutionMode, VmInterfaceExt};
+
+use super::{
+    read_validation_test_contract, tester::VmTesterBuilder, ContractToDeploy, TestedVm,
+    TestedVmForValidation,
+};
+use crate::interface::TxExecutionMode;
+
+/// Checks that every limitation imposed on account validation results in an appropriate error.
+/// The actual misbehaviours are found in "validation-rule-breaker.sol".
+pub(crate) fn test_account_validation_rules<VM: TestedVm + TestedVmForValidation>() {
+    assert_matches!(test_rule::<VM>(0), None);
+    assert_matches!(
+        test_rule::<VM>(1),
+        Some(ViolatedValidationRule::TouchedDisallowedStorageSlots(_, _))
+    );
+    assert_matches!(
+        test_rule::<VM>(2),
+        Some(ViolatedValidationRule::CalledContractWithNoCode(_))
+    );
+
+    // Disallowing gasleft is difficult because it isn't immediately followed
+    // by far call in EraVM bytecode.
+    /*assert_matches!(
+        test_rule::<VM>(100),
+        Some(ViolatedValidationRule::TouchedDisallowedContext)
+    );*/
+
+    // TODO: test running out of gas but catching the failure.
+    // Can be accomplished via many nested far calls.
+}
+
+fn test_rule<VM: TestedVm + TestedVmForValidation>(rule: u32) -> Option<ViolatedValidationRule> {
+    let aa_address = Address::repeat_byte(0x10);
+    let beneficiary_address = Address::repeat_byte(0x20);
+
+    let (bytecode, contract) = read_validation_test_contract();
+    let mut vm = VmTesterBuilder::new()
+        .with_empty_in_memory_storage()
+        .with_custom_contracts(vec![
+            ContractToDeploy::account(bytecode, aa_address).funded()
+        ])
+        .with_execution_mode(TxExecutionMode::VerifyExecute)
+        .with_rich_accounts(1)
+        .build::<VM>();
+
+    let mut private_account = vm.rich_accounts[0].clone();
+
+    // Set the type of misbehaviour of the AA contract
+    let function = contract.function("setTypeOfRuleBreak").unwrap();
```

Comment:
Nit: IMO, it would be acceptable to set the contract storage directly in this case to simplify the contract, but since this simplification is marginal, feel free to leave as-is.

Reply:
That's an interesting idea. I guess the number is always at slot 0, so it would be easy.